liberapay/models/community.py

```diff
@@ -38,8 +39,17 @@ def create(cls, name, creator_id, lang='mul'):
         name = unicodedata.normalize('NFKC', name)
         if name_re.match(name) is None:
             raise InvalidCommunityName(name)
+
         try:
             with cls.db.get_cursor() as cursor:
+                all_names = cursor.all("""
+                    SELECT name
+                    FROM communities
+                    """)
+                for existing_name in all_names:
+                    if cls._unconfusable(name) == cls._unconfusable(existing_name):
+                        raise CommunityAlreadyExists
+
```

Comment:
This is a very inefficient implementation, the more communities there are the slower it becomes. Also you're repeating an identical function call (`cls._unconfusable(name)`) in every loop iteration.

Reply:
Fixed here: 9b496bc88f8a3c404394ee8c1999e33a6f944a0b

Reply:
Which part is the most inefficient:

- the database query?
- the unconfusable lookup?
- the `unconfusable_string` function?

Reply:
The fetching of all the existing community names from the database and the subsequent loop. That's linear complexity, [O(n)](https://en.wikipedia.org/wiki/Big_O_notation), with `n` being the number of existing communities.

Reply:
We should probably do something like this instead:

```python
                possible_collisions = get_confusables(name)
                possible_collisions.append(name)
                collides_with = cursor.one("""
                    SELECT name
                      FROM communities
                     WHERE name IN %s
                     LIMIT 1
                """, (tuple(possible_collisions),))
                if collides_with:
                    raise CommunityAlreadyExists(collides_with)
```